lldb/packages/Python/lldbsuite/test/dotest.py

```diff
@@ -914,6 +914,25 @@ def checkForkVForkSupport():
         configuration.skip_categories.append("fork")
 
 
+def checkPexpectSupport():
+    from lldbsuite.test import lldbplatformutil
+
+    platform = lldbplatformutil.getPlatform()
+
+    # llvm.org/pr22274: need a pexpect replacement for windows
+    if platform in ["windows"]:
+        if configuration.verbose:
+            print("pexpect tests will be skipped because of unsupported platform")
+        configuration.skip_categories.append("pexpect")
+    elif not configuration.shouldSkipBecauseOfCategories(["pexpect"]):
+        try:
+            import pexpect
+        except:
+            print(
+                "Warning: pexpect is not installed, but pexpect tests are not being skipped."
```

Comment:
Note: instead of just printing a warning, we could make this also add `pexpect` to `skip_categories` as above. However, that would lead to a false notion that `pexpect` tests are running and passing. Someone who is running these tests may not be aware that `pexpect` needs to be installed, or maybe they are aware but a system change made them go away. IMHO, it is safer to let the tests fail and make the user acknowledge that `pexpect` tests should be skipped.

Reply:
For a pexpect test, it would fail either way, but raising an error here would be more clear about the failure.

For a non-pexpect test, this will unnecessarily fail. `ninja check-lldb-api` will have ~1k test failures instead of ~30 failures.

Updated to `raise` instead here.

Reply:
I'm confused, when will this run? I think you're saying that when a test in the pexpect category is run you'll get:
FAILED: "pexpect is not installed, but pexpect...."

Something like that. So you can still run tests from other categories, and it doesn't just stop right at the start of `check-lldb` and refuse to run anything.

Reply:
`checkPexpectSupport()` will be called for every test, just like the other check* methods above.

This `try` block here will only run if a non-Windows platform is configured to run `pexpect` tests, either because:
* The test is running w/ various `--skip-category=foo --skip-category=bar`, and `pexpect` is not one of the skipped categories, or
* The test is running w/ `--category=pexpect` (this is unusual)

I don't think there's a way to only run this check if the test has a certain category. We don't know what the category is, because the test file isn't loaded, and loading it would mean potentially evaluating the `import pexpect` statement that will fail.

The whole `elif` block is just to provide a better message if `pexpect` is not installed. If I remove it, then running a `pexpect` test will fail like so:

```
$ bin/lldb-dotest -p TestSTTYBeforeAndAfter.py
...
lldb version 19.0.0git (https://github.com/llvm/llvm-project.git revision b5e04eb49c89c8c654305c6ba5db5e2f237bccec)
  clang revision b5e04eb49c89c8c654305c6ba5db5e2f237bccec
  llvm revision b5e04eb49c89c8c654305c6ba5db5e2f237bccec
Skipping the following test categories: ['dsym', 'gmodules', 'debugserver', 'objc']
FAIL: LLDB (/home/rupprecht/src/llvm-build/dev/bin/clang-x86_64) :: test_stty_dash_a_before_and_afetr_invoking_lldb_command (TestSTTYBeforeAndAfter.TestSTTYBeforeAndAfter.test_stty_dash_a_before_and_afetr_invoking_lldb_command)
======================================================================
ERROR: test_stty_dash_a_before_and_afetr_invoking_lldb_command (TestSTTYBeforeAndAfter.TestSTTYBeforeAndAfter.test_stty_dash_a_before_and_afetr_invoking_lldb_command)
   Test that 'stty -a' displays the same output before and after running the lldb command.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/rupprecht/src/llvm-project/lldb/packages/Python/lldbsuite/test/decorators.py", line 446, in wrapper
    return func(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rupprecht/src/llvm-project/lldb/test/API/terminal/TestSTTYBeforeAndAfter.py", line 26, in test_stty_dash_a_before_and_afetr_invoking_lldb_command
    import pexpect
ModuleNotFoundError: No module named 'pexpect'
Config=x86_64-/home/rupprecht/src/llvm-build/dev/bin/clang
----------------------------------------------------------------------
Ran 1 test in 0.011s

FAILED (errors=1)
```

I suppose the error raised by the `elif` block, as written, doesn't provide much benefit on top of that -- it's clear that the issue is trying to run a pexpect test w/o pexpect being installed. So maybe I should replace it with something more actionable, e.g. to suggest configuring cmake w/ `-DLLDB_TEST_USER_ARGS=--skip-category=pexpect`. Or just remove it altogether?

Reply:
I understand now.

Yeah adding something actionable would be a good improvement, tell them to install pexpect or to skip the category.

Reply:
So what would happens without this code? Wouldn't we try to import pexpect anyway in the relevant test? In other words, do we even need this? It seems that printing a warning for non-pexpect either adds noise to the already verbose test output or will be invisible if the test passes. 

If the error message is poor enough that we would want to print something more actionable, could we do it from the `PExpectTest` base class?

Reply:
Proposed some more specific wording in https://github.com/llvm/llvm-project/pull/84860/commits/bcfc5e8f2f18e8eb8f5a96e18c649f3cb2e248c0

But this `elif` block is the least thing that I care about actually landing :) I think it's helpful, but if this is contentious, it's not a big deal to drop it and just have a potentially worse error message. So I dropped it altogether in https://github.com/llvm/llvm-project/pull/84860/commits/03e13152eac8416c31414b4f469e281c40b80deb. I can always add it back if needed, either in this PR or as a followup commit.

> So what would happens without this code? Wouldn't we try to import pexpect anyway in the relevant test? In other words, do we even need this?

Yep, see the snippet above ([link](#discussion_r1521684820)) for what it'd look like if we just remove it. Having a specific check+error text here is not essential, but might help some developer if they aren't aware of how to skip this category.

> It seems that printing a warning for non-pexpect either adds noise to the already verbose test output or will be invisible if the test passes.

Agreed about spamminess; that was changed in https://github.com/llvm/llvm-project/pull/84860/commits/7f2ec70a35d1f5426afcf354f9b16b0052e81df2 to make this a hard error.

> If the error message is poor enough that we would want to print something more actionable, could we do it from the PExpectTest base class?

That would not work for tests that use pexpect outside of `PExpectTest`, which are:
- lldb/test/API/terminal/TestSTTYBeforeAndAfter.py
- lldb/test/API/macosx/nslog/TestDarwinNSLogOutput.py
- lldb/test/API/benchmarks/expression/TestRepeatedExprs.py (and many other benchmark tests)

So if the user just runs `ninja check-lldb`, they'd get an actionable message from PExpectTest-based tests, but a potentially confusing one from TestSTTYBeforeAndAfter.